plugins/extract/extract.plugin.zsh

```diff
@@ -29,61 +29,75 @@ EOF
-      (*.xz) unxz "$file" ;;
-      (*.lrz) (( $+commands[lrunzip] )) && lrunzip "$file" ;;
-      (*.lz4) lz4 -d "$file" ;;
-      (*.lzma) unlzma "$file" ;;
-      (*.z) uncompress "$file" ;;
-      (*.zip|*.war|*.jar|*.ear|*.sublime-package|*.ipa|*.ipsw|*.xpi|*.apk|*.aar|*.whl) unzip "$file" -d "$extract_dir" ;;
-      (*.rar) unrar x -ad "$file" ;;
+        && tar --zstd -xvf "$full_path" \
+        || zstdcat "$full_path" | tar xvf - ;;
+      (*.tar) tar xvf "$full_path" ;;
+      (*.tar.lz) (( $+commands[lzip] )) && tar xvf "$full_path" ;;
+      (*.tar.lz4) lz4 -c -d "$full_path" | tar xvf - ;;
+      (*.tar.lrz) (( $+commands[lrzuntar] )) && lrzuntar "$full_path" ;;
+      (*.gz) (( $+commands[pigz] )) && pigz -dk "$full_path" || gunzip -k "$full_path" ;;
+      (*.bz2) bunzip2 "$full_path" ;;
+      (*.xz) unxz "$full_path" ;;
+      (*.lrz) (( $+commands[lrunzip] )) && lrunzip "$full_path" ;;
+      (*.lz4) lz4 -d "$full_path" ;;
+      (*.lzma) unlzma "$full_path" ;;
+      (*.z) uncompress "$full_path" ;;
+      (*.zip|*.war|*.jar|*.ear|*.sublime-package|*.ipa|*.ipsw|*.xpi|*.apk|*.aar|*.whl) unzip "$full_path" ;;
+      (*.rar) unrar x -ad "$full_path" ;;
       (*.rpm)
-        command mkdir -p "$extract_dir" && builtin cd -q "$extract_dir" \
-        && rpm2cpio "$full_path" | cpio --quiet -id ;;
-      (*.7z) 7za x "$file" ;;
+        rpm2cpio "$full_path" | cpio --quiet -id ;;
+      (*.7z) 7za x "$full_path" ;;
       (*.deb)
-        command mkdir -p "$extract_dir/control" "$extract_dir/data"
-        builtin cd -q "$extract_dir"; ar vx "$full_path" > /dev/null
+        command mkdir -p "control" "data"
+        ar vx "$full_path" > /dev/null
         builtin cd -q control; extract ../control.tar.*
         builtin cd -q ../data; extract ../data.tar.*
         builtin cd -q ..; command rm *.tar.* debian-binary ;;
-      (*.zst) unzstd "$file" ;;
-      (*.cab) cabextract -d "$extract_dir" "$file" ;;
-      (*.cpio|*.obscpio) cpio -idmvF "$file" ;;
-      (*.zpaq) zpaq x "$file" ;;
+      (*.zst) unzstd "$full_path" ;;
+      (*.cab) cabextract "$full_path" ;;
+      (*.cpio|*.obscpio) cpio -idmvF "$full_path" ;;
+      (*.zpaq) zpaq x "$full_path" ;;
       (*)
         echo "extract: '$file' cannot be extracted" >&2
         success=1 ;;
     esac
 
     (( success = success > 0 ? success : $? ))
-    (( success == 0 && remove_archive == 0 )) && rm "$full_path"
+    (( success == 0 && remove_archive == 0 )) && command rm "$full_path"
     shift
 
-    # Go back to original working directory in case we ran cd previously
+    # Go back to original working directory
+    # and remove extraction directory if there was an error
     builtin cd -q "$pwd"
+    (( success > 0 )) && command rm -r "$extract_dir"
+
+    # If content of extract dir is a single directory, move its contents up
+    # Glob flags: N: no error if directory is empty, Y2: at most 2 files
+    local content=("${extract_dir}"/*(NY2))
```

Comment:
Being a local array, shouldn't be declared and then initialized, or that is just with associative arrays?